colcon_bundle/verb/_archive_generators.py

```diff
@@ -0,0 +1,174 @@
+    """
+    # install_base: Directory with built artifacts from the workspace
+    install_base = path_context.install_base()
+    # staging_path: Directory where all dependencies have been installed
+    staging_path = path_context.staging_path()
+
+    logger.info('Copying {} into bundle...'.format(install_base))
+    assets_directory = os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), 'assets')
+    shellscript_path = os.path.join(assets_directory, 'v1_setup.sh')
+    shutil.copy2(shellscript_path, os.path.join(staging_path, 'setup.sh'))
+    bundle_workspace_install_path = os.path.join(
+        staging_path, 'opt', 'install')
+    if os.path.exists(bundle_workspace_install_path):
+        shutil.rmtree(bundle_workspace_install_path)
+    shutil.copytree(install_base, bundle_workspace_install_path)
+
+    logger.info('Archiving the bundle output')
+    print('Creating bundle archive...')
+
+    bundle_tar_path = path_context.bundle_tar_path()
+    metadata_tar_path = path_context.metadata_tar_path()
+    archive_tar_gz_path = path_context.archive_tar_gz_path()
+
+    with tarfile.open(metadata_tar_path, 'w') as archive:
+        archive.add(path_context.installer_metadata_path(),
+                    arcname='installers.json')
+
+    if os.path.exists(bundle_tar_path):
+        os.remove(bundle_tar_path)
+
+    _recursive_tar_in_path(bundle_tar_path, staging_path)
+
+    version_file_path = path_context.version_file_path()
+    with open(version_file_path, 'w') as v:
+        v.write('1')
+
+    with tarfile.open(
+            archive_tar_gz_path, 'w:gz', compresslevel=5) as archive:
+        archive.add(
+            version_file_path, arcname='version')
+        archive.add(
+            metadata_tar_path, arcname=os.path.basename(metadata_tar_path))
+        archive.add(
+            bundle_tar_path, arcname=os.path.basename(bundle_tar_path))
+
+    os.remove(metadata_tar_path)
+    os.remove(bundle_tar_path)
+
+    logger.info('Archiving complete')
+
+
+def generate_archive_v2(path_context,
+                        metadata_paths,
+                        dependencies_changed):
+    """
+    Generate bundle archive v2.
+
+    This archive is a tarfile that contains multiple compressed archives:
+
+    output.tar
+    |- version
+    |- metadata.tar.gz
+    |- pad (optional)
+    |- dependencies.tar.gz
+    |- workspace.tar.gz
+
+    :param path_context: PathContext object including all path configurations
+    :param metadata_paths: [str] paths to files which should be included
+    in the metadata archive
+    :param dependencies_changed: Boolean representing whether the staging path
+    needs to be re-archvied
+    """
+    logger.info('Archiving the bundle output')
+    print('Creating bundle archive V2...')
+    logger.debug('Start: workspace.tar.gz')
+    archive_tar_path = path_context.archive_tar_path()
+    workspace_tar_gz_path = path_context.workspace_tar_gz_path()
+
+    # Install directory
```

Comment:
This function is super long:

This pattern:
```py
def f():
  # something
  a()
  b()
  c()

  # something else
  d()
  e()

```

...is a code smell and means we should usually have instead:

```
def f1():
  """something""
  a()
  b()
  c()

def f2():
  """something else"""
  d()
  e()

def f():
  f1()
  f2()